src/dbnode/encoding/proto/custom_unmarshal.go

```diff
@@ -0,0 +1,386 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+)
+
+var (
+	// Groups in the Protobuf wire format are deprecated, so simplify the code significantly by
+	// not supporting them.
+	errGroupsAreNotSupported = errors.New("use of groups in proto wire format is not supported")
+	zeroValue                unmarshalValue
+)
+
+type customFieldUnmarshaler interface {
+	sortedCustomFieldValues() sortedCustomFieldValues
+	nonCustomFieldValues() *dynamic.Message
+	numNonCustomValues() int
+	resetAndUnmarshal(schema *desc.MessageDescriptor, buf []byte) error
+}
+
+type customUnmarshaler struct {
+	schema          *desc.MessageDescriptor
+	decodeBuf       *buffer
+	customValues    sortedCustomFieldValues
+	nonCustomValues *dynamic.Message
+	numNonCustom    int
+}
+
+func newCustomFieldUnmarshaler() customFieldUnmarshaler {
+	return &customUnmarshaler{
+		decodeBuf: newCodedBuffer(nil),
+	}
+}
+
+func (u *customUnmarshaler) sortedCustomFieldValues() sortedCustomFieldValues {
+	return u.customValues
+}
+
+func (u *customUnmarshaler) numNonCustomValues() int {
+	return u.numNonCustom
+}
+
+func (u *customUnmarshaler) nonCustomFieldValues() *dynamic.Message {
+	if u.nonCustomValues == nil {
+		u.nonCustomValues = dynamic.NewMessage(u.schema)
+	}
+	return u.nonCustomValues
+}
+
+func (u *customUnmarshaler) unmarshal() error {
+	u.customValues = u.customValues[:0]
+
+	if u.nonCustomValues != nil {
+		u.nonCustomValues.Reset()
+	}
+
+	isSorted := true
+	for !u.decodeBuf.eof() {
+		tagAndWireTypeStartOffset := u.decodeBuf.index
+		fieldNum, wireType, err := u.decodeBuf.decodeTagAndWireType()
+		if err != nil {
+			return err
+		}
+
+		fd := u.schema.FindFieldByNumber(fieldNum)
+		if fd == nil {
+			return fmt.Errorf("encountered unknown field with field number: %d", fieldNum)
+		}
+
+		if !u.isCustomField(fd) {
+			if u.nonCustomValues == nil {
+				u.nonCustomValues = dynamic.NewMessage(u.schema)
+			}
```

Comment:
Perhaps think about extracting this into a sub method in the future?

i.e.

```go
func (u *customUnmarshaler) nonCustomVals() *dynamic.Message {
  if u.nonCustomValues == nil {
    u.nonCustomValues = dynamic.NewMessage(u.schema)
  }
  return u.nonCustomValues
}
```

Then use as 

```go
// reset
u.nonCustomVals().Reset()
```

etc

Reply:
(it'll very likely get inlined)